### ISSUE TITLE
Feature request: fish_user_paths and syncing fish_variables

### ISSUE THREAD

**Issue by thernstig** [fish-shell/fish-shell#6641]
Note: I am not sure this is the best feature request, but I've got a legitimate problem I am not sure how to solve in a good way. In case it's not a good feature request, at least some future user can possibly find this issue and see the answer.

---

Some tools, like https://github.com/jorgebucaran/fish-nvm alter `fish_user_paths` at times. E.g. changing a node js version with `nvm use 10.15.13` will update my fish_user_paths like below:

```fish
SETUVAR fish_user_paths:/home/userA/\x2econfig/nvm/10\x2e15\x2e3/bin
```

Other tools also use universal variables e.g. fzf:

```fish
SETUVAR FZF_PREVIEW_FILE_CMD:head\x20\x2dn\x2010
```

Both are valid use cases and both alter `fish_variabes`. As time goes, a user is bound to have a lot of custom setup in fish_variables. So much, that they might want to sync all fish config between machines. E.g. home computer and work computer. I do this myself by creating a git repo out of my `~/.config/fish`. I think this is a valid use case, even editors like VS Code are now adding cloud sync capabilities since users move between machines. Syncing of user setup is a growing trend. I.e. no more re-install everything from scratch when getting a new machine. The point is - syncing setup/config is a valid use case.

Due to that `fish_user_paths` adds home directories with their full path, this will break across machines.  We cannot force fish packages to not use `fish_user_paths` either. Therefor I am asking if there is a possibility to add some feature like a `fish_user_paths` that uses tilde `~` in the path (this would most cases) or even better, possibility to have fish_variables per machine.

If I missed some obvious way to achieve this. please enligten me and we can close this feature request 😄

**Comment by krobelus** [comment]
I'd not sync `fish_variables` but create and sync a function that initializes any universal variables + abbreviations you need, and run that whenever you need to. I think of `fish_variables` as a cache, not as configuration.

Alternatively you could write a script to fix the paths in your universal variables after syncing (see the output of `set -U`).

**Comment by zanchey** [comment]
I think it would be reasonable to teach path variables about tilde expansion.

**Comment by thernstig** [comment]
@krobelus The problem with that approach is that you would never know when you edit/update/configure some installed package, i.e. via fisher, if it updates the `fish_variables` file. If you don't notice this, there is no chance to know that you need to update your sync function. When I have version controlled it, I can easily go into `~/.config/fish`, do a `git status` and see there has been changes. Even though I do config separately on two machines, later pushing/rebasing will let me discover this.

**Comment by ridiculousfish** [comment]
Keeping fish functions and config in git is very interesting - I think that should be a focus.

Teaching path variables about ~ seems reasonable. Bash does it:

    export $PATH "$PATH:~/abcd"

now $PATH contains a literal `~` but it gets expanded to your home directory. We could have a home-directory-only-expansion-mode.

Amusingly bash is inconsistent about other forms of expansion. If you have a literal `$VAR` in $PATH, then it will expand for tab completion but won't actually run it.

Expanding variables in variables is no good (what if it's cyclic, hah!), but tilde can probably be expanded.

**Comment by thernstig** [comment]
Since fish is a _a command line shell for the 90s_, a cool long-term vision would be to be able to sync all aspects of fish between machines and various sync providers (If you are VS Code user you might know what I mean). Imagine having seamless sync to the cloud provider of your choice and then it "just works" across machines. It'd almost be like the fish shell could be bumped to a _command line shell for the millennials_ 😮 . But this is off topic for this particular request.

**Comment by zanchey** [comment]
At least in my testing with bash 4.4, the tilde is expanded when the variable is set (`PATH=$PATH:~/foo; echo $PATH` gives `...:/home/zanchey/foo`), which I thinks makes sense - otherwise expansion has to happen when the variable is exported.

**Comment by thernstig** [comment]
@zanchey Yes, that's the problem that once set in `fish_user_paths` it's already expanded, so it's impossible to share the config in `fish_variables`

**Comment by danielb2** [comment]
how can I set `SETUVAR fish_color_autosuggestion:purple` manually in my config.fish file ? Since i don't have fish_variables checked into git, I would like to keep it seomwhere. I noticed on my FreeBSD machine the autosuggestion color is totally unreadable

**Comment by ridiculousfish** [comment]
Some fun discussion of this feature in bash, turns out it's not especially well liked:

https://www.mail-archive.com/bug-bash@gnu.org/msg19769.html
https://lists.gnu.org/archive/html/bug-bash/2014-07/msg00022.html

Remarkably it seems to be completely undocumented in bash and "rediscovered" repeatedly. [Here's the responsible line](https://github.com/bminor/bash/blob/d233b485e83c3a784b803fb894280773f16f2deb/findcmd.c#L527) and of course it depends on `POSIXLY_CORRECT`.

**Comment by thernstig** [comment]
Although that thread at https://www.mail-archive.com/bug-bash@gnu.org/msg19769.html states:

> While it's come up a couple of times in 28 years, it's not been a huge
problem.

It might be an indication that if this is changed, there would be an even larger influx of users hating that new change. I'm obviously quite pro allowing tilde expansion 😄 

Is this maybe a better thing that could be allowed? I tested it but it did not work:
```sh
# In the file fish_variables
SETUVAR TEST:{$HOME}\moo

# In the shell
~/.config/fish> echo $TEST
{$HOME}moo
~/.config/fish> echo {$TEST}
{$HOME}moo
```

Could references to other vars be expanded instead of tilde?

**Comment by thernstig** [comment]
With the new `fish_add_path` this has one again cropped as something I want 😄 I want to add this to my `config.fish`:

```fish
fish_add_path ~/.local/bin
```

I cannot do this, since it modifies `fish_user_paths` and changes it to include `home/user/.local/bin`. Then it breaks my entire dotfiles setup (to sync my config too multiple machines: work, home, remote hosts) etc.

I am honestly surprised this has not received more upvotes, there surely must be more fish and dotfile users than me who sync `config.fish`?

**Comment by faho** [comment]
>I cannot do this, since it modifies fish_user_paths

@thernstig 

You can, and it doesn't have to modify fish_user_paths.

First of all if you do add it to config.fish, it will expand the `~` *on that system*, so it should add the correct path.

Or you can choose to forego $fish_user_paths entirely and use `fish_add_path --path`, which makes it add to $PATH directly.

The problem only occurs if you decide to sync universal variables between systems *and* have some that include ~ or $HOME somewhere. That's a problem with universal variables (which, in their current form, are probably going to be removed). Simply use global variables for those bits and you're golden.

**Comment by thernstig** [comment]
@faho I sync my `fish_variables` as well. I should have been more explicit (thought me having created the topic made it implicit 😝). I agree with you, you are right this is not really a problem with `fish_user_paths` then in this scenario as long as I can use `--path`. I opted to use `fish_add_path -g` though as it worked. I will resort to that, good tips thanks! Also interesting to know unviersal vars might go away in their current form, was not aware of that.

**Comment by gerrywastaken** [comment]
@faho 
> First of all if you do add it to config.fish, it will expand the ~ on that system, so it should add the correct path.
It is the correct path.. but only on that system.
On linux the correct path might be:
/home/username
But on MacOS it's:
/Users/username

If you are syncing this using something like git, then you end up with a diff on one system or the other when you haven't changed anything.

If you treat it like a cache as @krobelus suggests and don't sync that file, then you end no longer being able to sync abbreviations added via `abbr --add`.
